---
title: "PhenoPath on culturing cells"
author: "Zhiyuan Hu"
date: "10/10/2018"
output: html_document
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, fig.width = 6, fig.height = 4)

library(scater)
library(tidyverse)
library(magrittr)
library(matrixStats)
library(ggrepel)
library(cowplot)
library(viridis)
library(splines)
library(DT)
library(limma)
library(edgeR)
library(RColorBrewer)
# source("https://bioconductor.org/biocLite.R")
# biocLite("goseq")
# biocLite("phenopath")
library(goseq)
library(phenopath)
library(ggbeeswarm)


# http://stackoverflow.com/questions/18509527/first-letter-to-upper-case
firstup <- function(x) {
   substr(x, 1, 1) <- toupper(substr(x, 1, 1))
x
}

set.seed(123L)

scale_vec <- function(x) (x - mean(x)) / sd(x)
```

Thank Kieran for providing this code kindly.

# Fit the CLVM model

```{r}
secretory <- readRDS("../rds/20180910Fresh_secretory_5_7clusters_clincluster.rds")
sce <- readRDS("../rds/20181011_culture_11553n15072_sceset.rds")
m <- model.matrix(~ sce$total_features + sce$Patient2)
# sce<- scater::normalise(sce)
sce<- scater::normaliseExprs(sce, design = m)
exprs(sce) <- norm_exprs(sce)

gene_vars <- rowVars(exprs(sce))
var_cutoff <- sort(gene_vars, decreasing = TRUE)[7500]
sce_hvg <- sce[gene_vars >= var_cutoff, ]

is_na_mgi_symbol <- is.na(rowData(sce_hvg)$mgi_symbol)
rowData(sce_hvg)$mgi_symbol <- rowData(sce_hvg)$Geneid

sce_hvg$time <- NA
sce_hvg$time[sce$source == "Fresh"] <- 0
sce_hvg$time[sce$source == "O.N."] <- 0.5
sce_hvg$time[sce$source == "Long" & sce$Patient2 == 11553] <- 2
sce_hvg$time[sce$source == "Long" & sce$Patient2 == 15072] <- 6

sce$time <- sce_hvg$time
```

And perform inference:

```{r}
pc1 <- prcomp(t(exprs(sce_hvg)), scale = TRUE)$x[,1]
pc1 <- scale_vec(pc1)
time_numeric <- as.numeric( sce$time)
pc1 <- pc1 * sign(cor(pc1, time_numeric))
x <- sce_hvg$time
# assay(sce, "norm_exprs") <- normalise(sce)

pcavi <- phenopath(sce_hvg, x, z_init = pc1, sce_assay = "norm_exprs")
```

Save for decision analysis:

```{r}
dec_df <- data_frame(gene = rownames(sce_hvg),
                     m = pcavi$m_beta[1,],
                     s = pcavi$s_beta[1,])
# write_csv(dec_df, "../tables/Phenopath/dec_df_new.csv")
```


Plot the elbo:

```{r plot-elbo}
plot_elbo(pcavi) +
  xlab(expression("CAVI iterations")) +
  theme_cowplot(font_size = 11)

ggsave("../plots/20181012phenopath/elbo.png", width = 5, height = 3)
```

# Exploring PhenoPath results

## Comparison to PC1

```{r pseudotime-vs-pc1}
xs <- sce_hvg$time
 
qplot(scale_vec(pc1), pcavi$m_z, color = factor(xs)) +
  scale_color_brewer(palette = "Set1", name = "Source") +
  #stat_function(fun = function(x) x, color = 'black') + 
  xlab("PC1") + ylab(expression(z))
```


```{r}
qplot(scale_vec(pc1), pcavi$m_z, color = factor(sce$Sample.Name)) +
  scale_color_brewer(palette = "Set1", name = "Source") +
  #stat_function(fun = function(x) x, color = 'black') + 
  xlab("PC1") + ylab(expression(z))
```


## Identifying significant genes


```{r beta-df}
df_beta <- data_frame(
  lambda = pcavi$m_lambda,
  beta = pcavi$m_beta[1,],
  chi = pcavi$chi_exp[1,],
  alpha = pcavi$m_alpha[1,],
  pos_sd = sqrt(pcavi$s_beta[1,]),
  gene = rowData(sce_hvg)$Geneid, #stringr::str_to_title(rownames(sce_hvg)),
  # ensembl_gene_id = rowData(sce_hvg)$ensembl_gene_id,
  is_sig = significant_interactions(pcavi, n = 3) #[,1]
)


# write_csv(df_beta, "../tables/Phenopath/df_gene_shalek.csv")


df_beta$is_sig_graph <- 
  plyr::mapvalues(df_beta$is_sig, from = c(FALSE, TRUE),
                  to = c("Non-significant", "Significant"))
```

This gives:

```{r how-many-sig}
sum(df_beta$is_sig)
```

_significant_ interactions.

We can plot these:

```{r plot-sigs}
chi_cutoff <- 0.15
df_beta <- dplyr::mutate(df_beta, 
                  gene_symbol = gene)


ggplot(df_beta, aes(x = beta, y = 1 / chi, color = is_sig_graph)) + 
  geom_point(alpha = 0.8) +
  theme(legend.title = element_blank()) +
  # geom_point(data = dplyr::filter(df_beta, is_mmr), alpha = 0.6) +
  scale_color_brewer(palette = "Set1", name = "'Significant'") +
     geom_label_repel(data = dplyr::filter(df_beta, 1 / chi > chi_cutoff), aes(label = gene_symbol)) +
  xlab(expression(beta)) +
  ylab(expression(paste("[", chi, "]"^-1)))
  
```

Which genes are sig?

```{r dt-sig}
dplyr::filter(df_beta, is_sig) %>% dplyr::arrange(desc(abs(beta))) %>% 
  dplyr::select(gene, beta, alpha) %>% 
  datatable()
```

We can plot expression along phenotime for the 35 _significant_ genes:

```{r exprs-along-pst, fig.width = 12, fig.height = 9}
tmap <- pcavi$m_z

top_genes <- dplyr::filter(df_beta, is_sig) %>% 
  dplyr::arrange(desc(abs(beta))) %>% 
  extract2("gene") %>% head(n=20)
```

Check z loosely corresponds to time:

```{r dfz, fig.height=2.5, fig.width=8}
zdf <- data_frame(z = pcavi$m_z, time = as.factor(sce_hvg$time), stimulant = xs, patient = sce_hvg$Patient)

ggplot(zdf, aes(x = time, y = z, fill = time)) + 
  geom_violin(alpha = 0.8) +
  theme(legend.position = "none") + 
  scale_fill_brewer(palette = "Set3") +
  # scale_fill_brewer(palette = "Set1") +
  xlab("Capture time") +
  ylab("Pseudotime)") +
     facet_grid(~patient, scales = "free")
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 9),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) 

# ggsave(filename = "../plots/20181012phenopath/violin_pseudotimeVSpatient.png",width = 8, height = 2 )
  
time_plot <- last_plot()
```

```{r dfz2, fig.height=2.5, fig.width=2}
zdf <- data_frame(z = pcavi$m_z, time = as.factor(sce_hvg$time), stimulant = xs, patient = sce_hvg$Patient)

ggplot(zdf, aes(x = time, y = z, fill = time)) + 
  geom_violin(alpha = 0.8) +
  theme(legend.position = "none") + 
  scale_fill_brewer(palette = "Set3") +
  # scale_fill_brewer(palette = "Set1") +
  xlab("Capture time (day)") +
  ylab("Pseudotime") +
     # facet_grid(~patient, scales = "free")
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 15),
        legend.title = element_text(size = 15),
        legend.text = element_text(size = 15)) 

# ggsave(filename = "../plots/20181012phenopath/violin_pseudotime.png",width = 3, height = 3)
# ggsave(filename = "../manuscript_plots/PhenoPath/violin_pseudotime.png",width = 3, height = 3)
  

# time_plot <- last_plot()
```

```{r, fig.height=5, fig.width=15}
select <- dplyr::select
sce_hvg <- runPCA(sce_hvg, ncomponents = 3, return_SCE = TRUE)

tmap <- pcavi$m_z
sce_hvg$tmap <- tmap

df_pca <- as_data_frame(reducedDim(sce_hvg)[,c(1,2)]) %>% 
  mutate(pseudotime = sce_hvg$tmap, Stimulant = sce_hvg$source)

df_no_stim <- select(df_pca, -Stimulant)

df_pca$Stimulant <- factor(df_pca$Stimulant, levels = c("Fresh","O.N.","Long"))

plt <- ggplot(df_pca, aes(x = PC1, y = PC2)) +
  geom_point(data = df_no_stim, fill = "grey80", color = "grey80", size = 3) +
  geom_point(aes(fill = pseudotime), shape = 21, color = 'grey20', size = 3) +
  scale_fill_viridis(name = "Pseudotime", option = "C") + 
  facet_wrap(~ Stimulant) +
  theme(legend.position = c(.4,.08),
        legend.direction = "horizontal") +
  theme(strip.background = element_rect(fill = "grey95", color = "grey30", size = 1),
        legend.text = element_text(size = 8),
        axis.text = element_text(size = 9),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 10)) +
  xlim(-18, 20) + ylim(-15, 10)

plt

# ggsave("../plots/20181012phenopath/PC1PC2.png", width = 15,height = 5)
```

```{r}
tiff
```


```{r}
library(grid)
genes <- c("FOS", "JUN", "CCND1")
mm <- match(genes, rowData(sce_hvg)$mgi_symbol)
gexp <- t(exprs(sce_hvg)[mm, ])
colnames(gexp) <- genes
gexp_df <- as_data_frame(gexp) %>% 
  mutate(pseudotime = sce_hvg$tmap)

df_bar <- inner_join(df_pca, gexp_df)

classify_cell <- function(row) {
  if(row['pseudotime'] < 0.2) return("beginning")
  if(row['pseudotime'] > 0.2 && row['Stimulant'] == "O.N.") return("Overnight-end")
  if(row['pseudotime'] > 0.2 && row['Stimulant'] == "Long") return("Long-end")
  return(NA)
}

df_bar$cell_class <- sapply(seq_len(nrow(df_bar)), function(i) classify_cell(df_bar[i, ]))

df_group <- filter(df_bar, !is.na(cell_class)) %>% 
  gather(gene, expression, -(PC1:Stimulant), -cell_class) %>% 
  group_by(cell_class, gene) %>% 
  summarise(mean_expression = mean(expression))
df_group$gene <- factor(df_group$gene, levels = genes)

plts <- list()
for(cl in unique(df_group$cell_class)) {
  plts[[cl]] <- filter(df_group, cell_class == cl) %>% 
  ggplot(aes(x = gene, y = mean_expression, fill = gene)) +
    geom_bar(stat = "identity", color = 'grey20') +
    scale_fill_brewer(palette = "Dark2") +
    theme(legend.position = "None",
          axis.title.x = element_blank(),
          axis.title.y = element_text(size = 8, face = "bold"),
          axis.text.y = element_text(size = 8),
          axis.text.x = element_text(size = 8, angle = -90, face = "bold")) +
    labs(y = "Mean z-score\n expression") +
    ylim(-6, 6)
}

width = 0.12
height = 0.3

pca_plot <- ggdraw() + 
  draw_plot(plt) +
  draw_plot(plts[["beginning"]], x = 0.41, y = 0.2, width = width, height = height) +
  draw_plot(plts[["beginning"]], x = 0.88, y = 0.15, width = width, height = height) +
  draw_plot(plts[["Overnight-end"]], x = 0.06, y = 0.105, width = width, height = height) +
  draw_plot(plts[["Long-end"]], x = 0.56, y = 0.61, width = width, height = height)

pca_plot
```




```{r mlh1}

make_gene_plot <- function(genes) {
  inds <- match(genes, rowData(sce_hvg)$Geneid)
  
  df <- t(exprs(sce_hvg))[, inds, drop=FALSE] 
  colnames(df) <- genes #rownames(sce_hvg)[inds]
  
  df <- as_data_frame(df) %>% 
    dplyr::mutate(phenopath = pcavi$m_z, x = factor(sce_hvg$source,levels = c("Fresh","O.N.","Long"))) %>% 
    gather(gene, expression, -phenopath, -x)
  
  df$gene <- factor(df$gene, levels = genes)
  
  df_nox <- dplyr::select(df, -x)
  
  ggplot(df, aes(x = phenopath, y = expression, fill = x, color = x)) + 
    geom_point(data = df_nox, color = "grey80", fill = "grey80") +
    geom_point(shape = 21, color = 'grey30', alpha = 0.3) +
    facet_grid(gene ~ x, scales = "free") + 
    scale_fill_brewer(palette = "Set1", name = "MSI") +
    scale_color_brewer(palette = "Set1", name = "MSI") +
    theme(legend.position = "none", 
          strip.text = element_text(face = "bold"),
          strip.text.x = element_text(size = 10, margin = margin(.1, 0, .1, 0, "cm")),
          strip.text.y = element_text(size = 8, margin = margin(0, .1, 0, .1, "cm")),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.3, linetype = 1)) +
  theme(axis.text = element_text(size = 9),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 10)) +
    stat_smooth(se = FALSE, method = "lm", size = 1.5, formula = y ~ ns(x,3),
                color = 'grey30') +
    ylab(expression("Normalised expression")) + 
    xlab("Pseudotime")
}

genes <- c("ESR1","LGR5")

make_gene_plot(genes)
# ggsave("../manuscript_plots/PhenoPath/ESR1_LGR5.png", height = 3, width = 8)
# gene_plot <- plot_over_pseudotime <- last_plot()
```

Now do the same for original time series:

```{r}
inds <- match(genes, rowData(sce_hvg)$mgi_symbol)

df <- t(exprs(sce_hvg))[, inds, drop=FALSE] 
colnames(df) <- genes
  
df <- as_data_frame(df) %>% 
    dplyr::mutate(capture_time = sce_hvg$time, x = as.character(xs)) %>% 
    gather(gene, expression, -capture_time, -x)
  
df$gene <- factor(df$gene, levels = genes)

df_g <- group_by(df, capture_time, x, gene) %>% 
  summarise(median_expression = median(expression))

ggplot(df, aes(x = capture_time, y = expression, fill = x, color = x)) + 
    # geom_beeswarm(shape = 21, color = 'grey30', alpha = 0.3) +
    geom_jitter(shape = 21, color = 'grey30', alpha = 0.3, width= 0.1) +
    facet_grid(gene ~ x, scales = "free") + 
    scale_fill_brewer(palette = "Set1") +
    scale_color_brewer(palette = "Set1") +
    theme(legend.position = "none", 
          strip.text = element_text(face = "bold"),
          strip.text.x = element_text(size = 10, margin = margin(.1, 0, .1, 0, "cm")),
          strip.text.y = element_text(size = 8, margin = margin(0, .1, 0, .1, "cm")),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.3, linetype = 1)) +
  theme(axis.text = element_text(size = 9),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 10)) +
    ylab(expression("Normalised expression")) + 
    xlab("Capture time") +
      geom_point(data = df_g, aes(y = median_expression), 
                 shape = 21, color = 'black', size = 2) +
  geom_line(data = df_g, aes(y = median_expression, group = 1))

plot_over_capture_time <- last_plot()

# ggsave("../../figs/supplementary_time_shalek.png", width = 6, height = 4)
```


# Differential expression comparison

We'll use limma-voom for standard differential expression:


```{r limma}
dge <- DGEList(counts(sce_hvg))
dge <- calcNormFactors(dge)

design <- model.matrix(~ x, colData(sce_hvg))
v <- voom(dge, design, plot = TRUE)

fit <- lmFit(v, design)
fit <- eBayes(fit)
results <- decideTests(fit)


df_limma <- data_frame(coef = fit$coefficients[,2], 
                       pval = fit$p.value[,2],
                       mu = pcavi$m_mu,
                       gene_symbol = rowData(sce_hvg)$Geneid) %>% 
  left_join(df_beta, by = "gene_symbol") %>%
  dplyr::mutate(qval = p.adjust(pval, method = "BH"),
                log10qval = -log10(qval))
```


```{r compare-beta}

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

ggplot(df_limma, aes(x = beta, y = log10qval, color = is_sig_graph)) + 
  geom_point() +
  ylab(expression(paste("Limma Voom -", log[10], "(q-value)"))) + 
  xlab(expression(paste("PhenoPath ", beta))) +
  geom_hline(yintercept = -log10(0.05), linetype = 2, alpha = 0.5) +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 9),
        axis.title.y = element_text(size = 8)) +
  theme(legend.title = element_text(size = 9),
        legend.text = element_text(size = 8)) +
  scale_color_manual(values = cols2, name = "Interaction") +
  theme(legend.margin = margin(), legend.position = "none")

de_plot <- ggExtra::ggMarginal(last_plot(), margins = "y", type = "histogram", size = 10)
```

Again no obvious relationship.

# Crossover bit for chris

```{r df-beta}
df_beta <- mutate(df_beta,
                  crossover = - alpha / beta)

filter(df_beta, is_sig) %>% 
  ggplot(aes(x = crossover)) + geom_histogram()
```


```{r save-supp-data}
# write_csv(df_beta, "../tables/Phenopath/shalek_interactions.csv")
```

```{r more-vis}
textinfo <- frame_data(
  ~x, ~y, ~label,
  1.2, 0.2, "Gene upregulated along pseudotime\nLPS increases upregulation",
  1.2, -0.2, "Gene upregulated along pseudotime\nPAM increases upregulation",
  -1.3, 0.2, "Gene downregulated along pseudotime\nPAM increases downregulation",
  -1.3, -0.2, "Gene downregulated along pseudotime\nLPS increases downregulation"
)

cols <- RColorBrewer::brewer.pal(3, "Set2")
cols2 <- c("#c5e2d9", cols[2])

outline_cols = c("#c5e2d9", 'black')

ggplot(df_beta, aes(x = lambda, y = beta)) +
  geom_point(shape = 21, aes(fill = is_sig_graph, color = is_sig_graph)) +
  geom_vline(xintercept = 0, linetype = 2, alpha = 0.5) +
  geom_hline(yintercept = 0, linetype = 2, alpha = 0.5) +
  scale_fill_manual(values = cols2, name = "Interaction") +
  scale_color_manual(values = outline_cols, name = "Interaction") +
  geom_text_repel(data = dplyr::filter(df_beta, (beta > 1 & is_sig) | beta < -1),
                aes(label = gene), color = 'black',
                size = 3) +
  ylab("Covariate-pseudotime interaction") +
  xlab("Gene regulation over pseudotime") +
  theme(legend.position = 'bottom',  #, legend.margin = margin(),
        axis.text = element_text(size = 10),
        axis.title = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10)) 
  # geom_text(data = textinfo, aes(x = x, y = y, label = label), 
            # color = 'black', size = 3, fontface = "bold")
lplot <- ggExtra::ggMarginal(last_plot(), margins = "y", type = "histogram", size = 10)
# 
# ggsave("../manuscript_plots/")
```


```{r}
print(lplot)
```

# GO analysis

```{r}
genome <- "hg19"
# supportedGeneIDs()
id <- "geneSymbol"
# source("https://bioconductor.org/biocLite.R")
# biocLite("org.Hs.eg.db")

all_genes <- rowData(sce_hvg)$Geneid

upreg_genes <- filter(df_beta, is_sig, beta > 0) %>% extract2("gene_symbol") 
downreg_genes <- filter(df_beta, is_sig, beta < 0) %>% extract2("gene_symbol")

upg <- 1 * (all_genes %in% upreg_genes)
downg <- 1 * (all_genes %in% downreg_genes)

names(upg) <- names(downg) <- all_genes

pwfup <- nullp(upg, genome = "hg19", id = "geneSymbol")
goup <- goseq(pwfup, genome = "hg19", id = "geneSymbol", test.cats = "GO:BP")

pwfdown <- nullp(downg, genome = "hg19", id = "geneSymbol")
godown <- goseq(pwfdown, genome = "hg19", id = "geneSymbol", test.cats = "GO:BP")
```


Graph results:

```{r}
parse_go <- function(go, type, n_tested) {
  go <- go %>% 
    mutate(qval = p.adjust(over_represented_pvalue)) %>% 
    mutate(log10qval = -log10(qval),
           prop_in_cat = numInCat / n_tested) %>% 
    head(n = 6) %>% 
    mutate(type = type) %>% 
    tbl_df() %>% 
    arrange(desc(log10qval))
  go
}


gos <- bind_rows(
  parse_go(goup, "Up-regulated", length(upreg_genes)),
  parse_go(godown, "Down-regulated", length(downreg_genes))
)

# gos <- filter(gos, type == "LPS up-regulated") 

gos$term <- stringr::str_to_title(gos$term)
gos$term <- factor(gos$term, levels = gos$term[order(gos$log10qval)])

gos <- mutate(gos, prop_cat = numDEInCat / numInCat)


gos %>% 
  filter(qval < 0.05) %>% 
ggplot(aes(x = term, y = log10qval)) + #, size = 100 * prop_cat)) +
  geom_segment(aes(y = min(log10qval - 1), yend = log10qval, x = term, xend = term),
               color = 'grey30', linetype = 3) +
    geom_point(aes(color = type, shape = type)) +
    coord_flip() +
  facet_wrap(~ type, ncol = 1, scales = "free") +
    theme(axis.title.y = element_blank()) +
    ylab(expression(paste("-", log[10], " q-value"))) +
  scale_size(name = "% category\nrepresented") +
    theme(legend.position = "bottom", #c(0.6,0.1),
          legend.direction = "horizontal",
        axis.text = element_text(size = 8),
        axis.text.y = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.title = element_blank(),
        legend.text = element_text(size = 10),
        legend.margin = margin(t = -.3, l = -4.5, unit = "cm")) +
  scale_color_brewer(palette = "Set1") +
  theme(strip.text = element_text(size = 8, face = "bold"),
          strip.background = element_rect(colour = "black", fill = "grey95", 
                                          size = 0.5, linetype = 1))
# ggsave("../../figs/go.png", width = 5, height = 5)
# goplot <- last_plot()
```



# Technical

```{r technical}
sessionInfo()
```

## R version 3.5.1 (2018-07-02)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
##  [1] grid      splines   stats4    parallel  stats     graphics  grDevices
##  [8] utils     datasets  methods   base     
## 
## other attached packages:
##  [1] org.Hs.eg.db_3.6.0          AnnotationDbi_1.42.1       
##  [3] bindrcpp_0.2.2              ggbeeswarm_0.6.0           
##  [5] phenopath_1.4.0             goseq_1.32.0               
##  [7] geneLenDataBase_1.16.0      BiasedUrn_1.07             
##  [9] RColorBrewer_1.1-2          edgeR_3.22.5               
## [11] limma_3.36.5                DT_0.4                     
## [13] viridis_0.5.1               viridisLite_0.3.0          
## [15] cowplot_0.9.3               ggrepel_0.8.0              
## [17] magrittr_1.5                forcats_0.3.0              
## [19] stringr_1.3.1               dplyr_0.7.6                
## [21] purrr_0.2.5                 readr_1.1.1                
## [23] tidyr_0.8.1                 tibble_1.4.2               
## [25] tidyverse_1.2.1             scater_1.8.4               
## [27] SingleCellExperiment_1.2.0  SummarizedExperiment_1.10.1
## [29] DelayedArray_0.6.6          BiocParallel_1.14.2        
## [31] matrixStats_0.54.0          GenomicRanges_1.32.7       
## [33] GenomeInfoDb_1.16.0         IRanges_2.14.12            
## [35] S4Vectors_0.18.3            ggplot2_3.0.0              
## [37] Biobase_2.40.0              BiocGenerics_0.26.0        
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_1.3-2         rjson_0.2.20            
##  [3] rprojroot_1.3-2          XVector_0.20.0          
##  [5] rstudioapi_0.8           bit64_0.9-7             
##  [7] lubridate_1.7.4          xml2_1.2.0              
##  [9] tximport_1.8.0           knitr_1.20              
## [11] jsonlite_1.5             Rsamtools_1.32.3        
## [13] broom_0.5.0              GO.db_3.6.0             
## [15] shinydashboard_0.7.0     shiny_1.1.0             
## [17] compiler_3.5.1           httr_1.3.1              
## [19] backports_1.1.2          assertthat_0.2.0        
## [21] Matrix_1.2-14            lazyeval_0.2.1          
## [23] cli_1.0.1                later_0.7.5             
## [25] prettyunits_1.0.2        htmltools_0.3.6         
## [27] tools_3.5.1              gtable_0.2.0            
## [29] glue_1.3.0               GenomeInfoDbData_1.1.0  
## [31] reshape2_1.4.3           Rcpp_0.12.19            
## [33] cellranger_1.1.0         Biostrings_2.48.0       
## [35] nlme_3.1-137             rtracklayer_1.40.6      
## [37] crosstalk_1.0.0          DelayedMatrixStats_1.2.0
## [39] rvest_0.3.2              miniUI_0.1.1.1          
## [41] mime_0.6                 XML_3.98-1.16           
## [43] zlibbioc_1.26.0          scales_1.0.0            
## [45] hms_0.4.2                promises_1.0.1          
## [47] rhdf5_2.24.0             yaml_2.2.0              
## [49] memoise_1.1.0            gridExtra_2.3           
## [51] biomaRt_2.36.1           ggExtra_0.8             
## [53] stringi_1.2.4            RSQLite_2.1.1           
## [55] GenomicFeatures_1.32.3   rlang_0.2.2             
## [57] pkgconfig_2.0.2          bitops_1.0-6            
## [59] evaluate_0.12            lattice_0.20-35         
## [61] Rhdf5lib_1.2.1           bindr_0.1.1             
## [63] labeling_0.3             GenomicAlignments_1.16.0
## [65] htmlwidgets_1.3          bit_1.1-14              
## [67] tidyselect_0.2.5         plyr_1.8.4              
## [69] R6_2.3.0                 DBI_1.0.0               
## [71] mgcv_1.8-24              pillar_1.3.0            
## [73] haven_1.1.2              withr_2.1.2             
## [75] RCurl_1.95-4.11          modelr_0.1.2            
## [77] crayon_1.3.4             rmarkdown_1.10          
## [79] progress_1.2.0           locfit_1.5-9.1          
## [81] readxl_1.1.0             data.table_1.11.8       
## [83] blob_1.1.1               digest_0.6.18           
## [85] xtable_1.8-3             httpuv_1.4.5            
## [87] munsell_0.5.0            beeswarm_0.2.3          
## [89] vipor_0.4.5
