---
title: "Deconvolution of bulk expression data"
author: "Zhiyuan Hu"
date: "2019/01/21"
output: html_document
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, fig.width = 6, fig.height = 4)

# devtools::install_github('shenorrlab/bseq-sc')
library(bseqsc)

library(MASS)
library(xbioc)
library(e1071) #svm
library(preprocessCore)

library(survminer)
library(survival)
library(powerSurvEpi)
library(survminer)

library(Biobase)
library(SingleCellExperiment)
library(scater)

library(ggbeeswarm)
library(ggrepel)
library(viridis)

library(edgeR)
library(limma)
# library(copynumber)

source("modified_functions20190204.R")
```

## Summary

## Single cell data

```{r read-single-cell-data}
# Single cell data - Fresh cells
secretory <- readRDS("../rds/20190120Fresh_secretory_9clusters_clincluster.rds")
sceset <- readRDS("../rds/20180917Sceset_12clusters.rds")
markers2 <- read.csv("../tables/20190120Clincluster_fresh_secretory_9clusters_markers.csv", row.names = 1, as.is = T)
markers_sceset <- read.csv("../tables/20180917sceset_markers_12cluster-Zhiyuan’s iMac.csv", row.names = 1, as.is = T)
fresh <- sceset[,sceset$source == "Fresh"]
# rm(secretory)
rm(sceset)
```

## Select Markers

```{r select-markers}
# markers2$metrics <- markers2$logFC*log1p((markers2$ratio1/markers2$ratio2))

marker.C3 <- markers2[markers2$cluster == "C3" & markers2$logFC > 1.79, ]
marker.C4 <- markers2[markers2$cluster == "C4" & markers2$logFC > 1.5 & markers2$ratio2 < 0.3,]
marker.EMT <- markers2[markers2$cluster == "C8" & markers2$logFC > 2 & markers2$ratio2 < 0.2,]
marker.C10 <- markers2[markers2$cluster == "C10" & markers2$logFC > 2.5 & markers2$ratio2 < 0.2,]
marker.ciliated <- markers_sceset[markers_sceset$cluster == 9 & markers_sceset$logFC > 6 & markers_sceset$ratio2 < 0.1,]

markers <- list(C3 = marker.C3$gene,
                C4 = marker.C4$gene,
                EMT = marker.EMT$gene,
                C10 = marker.C10$gene,
                ciliated = marker.ciliated$gene)
# saveRDS(ftMakers, "../rds/20190121FTEsc_signatures_BSEQ-sc_deconvolution.rds",compress = T)
str(markers)
```

#### Estimation of cell type proportions

```{r perpare-fresheset}
fresheset <- ExpressionSet(assayData = as.matrix(counts(fresh)),
                           phenoData = new("AnnotatedDataFrame",
                                           data=as.data.frame(fresh@colData)))
fresheset$cell.type <- NA
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C3"]] <- "C3"
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C4"]] <- "C4"
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C8"]] <- "EMT"
fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C10"]] <- "C10"
# fresheset$cell.type[fresheset$Sample %in% secretory$Sample[secretory$clincluster_final == "C9"]] <- "C9"
fresheset$cell.type[fresheset$clincluster == 9] <- "ciliated"
fresheset <- fresheset[,!is.na(fresheset$cell.type)]
table(fresheset$cell.type)
```

```{r prepare-fresh-SingleCellExperiment_object}
fresh <- fresh[,match(colnames(fresheset), colnames(fresh))]
fresh$cell.type <- fresheset$cell.type
```


## Data

The eset datasets are preprocessed by '20181028Prepare_datasets.rmd'

The signature matrix is actually the mean expression within one cluster after normalisation (non-log-space).

#### TCGA data 

```{r read-tcga-data}
## TCGA data
eset <- readRDS("rds/20181029TCGA_eset.rds")
exprs_tcga <- readRDS("rds/20181029TCGA_exprs_nonlog.rds")

tcga <- SingleCellExperiment(assays = list(counts = eset@assayData$exprs),
                             colData = as.data.frame(pData(eset)))
logcounts(tcga) <- log2(calculateCPM(tcga) + 1)
```


#### Tothill dataset

```{r read-tothill-data}
toteset <- readRDS("rds/20181029_tothill_eset.rds")
exprs_tot <- readRDS("rds/20181029_tothill_exprs_nonlog.rds")

pData(toteset)$OS <- FALSE
pData(toteset)$OS[pData(toteset)$patient.status == "D"] <- TRUE
```

#### OXO-PCR datasets

```{r}
oxopcr_pre <- readRDS("../../../OXO_PCR/R/rds/OXO_PCR_Allsamples_2018Oct_PreChemo.rds")
oxopcr_info_pre <- readRDS("../../../OXO_PCR/R/rds/OXO_PCR_Allsamples_2018Oct_PreChemo_sampleInfo.rds")

# oxopcr <- readRDS("rds/20190215OXO_PCR_pre_tumors_and1stroma.rds") 
# oxopcr_info <- readRDS("rds/20190215OXO_PCR_pre_tumors_and1stroma_sampleinfo.rds")
```


## Select genes by PCA in TCGA data

```{r marker_pca_function}
pca_emt <- function(markers){
  markers_in <-intersect(markers, rownames(tcga))
  exprs_mat <- logcounts(tcga)[match(markers_in, rownames(tcga)), ,drop = F]

  ## scale
  exprs_mat <- t(exprs_mat)
  vars <- DelayedMatrixStats::colVars(DelayedArray(exprs_mat))
  exprs_mat <- sweep(exprs_mat, MARGIN=2, # We scale by the standard deviation, which also changes the centre.
             STATS=sqrt(vars), FUN="/", 
             check.margin=FALSE) # divide by the square root of vars

  ## PCA
  pca <- prcomp(exprs_mat)

  emt_w <- pca$rotation[,1]
  emt_w <- emt_w[order(emt_w, decreasing = T)]
  
  emt_w 
}
```

```{r select-markers-by-pcs}
## genes with too good correlations can go (i.e. cor > 0.8)
# cor(t(log1p(exprs_tcga)[new_markers$EMT,]))

new_markers <- list()

# new_markers$progenitor <- markers$progenitor
out <- pca_emt(markers = markers$C3)
out 
new_markers$C3 <- names(out)[abs(out) > 0.2]

out <- pca_emt(markers = markers$C4)
out
new_markers$C4 <- names(out)[abs(out) > 0.2]

out <- pca_emt(markers = markers$EMT)
out
new_markers$EMT <- names(out)[abs(out) > 0.2]
# cor(t(log1p(exprs_tcga)[new_markers$EMT,]))

out <- pca_emt(markers = markers$C10)
out
new_markers$C10 <- names(out)[abs(out) > 0.2]

out <- pca_emt(markers = markers$ciliated)
out
new_markers$ciliated <- names(out)[abs(out) > 0.2]
new_markers$ciliated <- new_markers$ciliated[c(1,2,3,8,9,10,13,14)] # remove redundant geens by cor > 0.8
str(new_markers)
```

```{r length-of-signaure}
length(unlist(new_markers))
```

52 genes in total

## BSEQ-sc

```{r bseq-signature-matrix}
sig_matrix_new <- bseq_my(markers = new_markers)
## upper limit
quantile(sig_matrix_new, 0.99)
#      99% 
# 897.1952 
sig_matrix_new[sig_matrix_new > quantile(sig_matrix_new, 0.99)] <- quantile(sig_matrix_new, 0.99) # average the too high value
```

```{r plot-signature-matrix}
# plotBasis(sig_matrix_new, new_markers, Colv = NA, Rowv = NA, layout = '_', col = 'Blues')
```

```{r save-plot-and-data-about-signature-matrix}
# pdf("plots_explore/20190207/sig_matrix_new.pdf", width = 4, height = 4)
# plotBasis(sig_matrix_new, new_markers, Colv = NA, Rowv = NA, layout = '_', col = 'Blues')
# dev.off()

# saveRDS(sig_matrix_new, "rds/20190213sig_matrix_new.rds", compress = T)
# saveRDS(new_markers, "rds/20190213new_markers.rds", compress = T)
# saveRDS(fresheset, "rds/20190213fresheset.rds", compress = T)

# write.csv(sig_matrix_new,"20190204_selected_signature_matrix_52genes.csv")

# new_markers <- readRDS("rds/20190213new_markers.rds")
```


#### Cibersort TCGA

Use the updated matrix on TCGA data

```{r deconvolute-TCGA-data-and-heatmap}
svr_fit_new <- cibersort_my(sig_matrix = sig_matrix_new, exprs_unlogged = exprs_tcga)
eset$EMT <- svr_fit_new[,"EMT"]
head(svr_fit_new)
p <- pheatmap::pheatmap(svr_fit_new[,1:5],show_rownames = F, cluster_cols = F)
# p <- ggplotify::as.ggplot(p)
# ggsave(filename = "plots_explore/20190207/deconvolution_matrix_TCGA.pdf", device = "pdf", width = 3,height = 6)
p
# write.csv(svr_fit_new[,1:5], "results/20190423TCGA_deconvolution.csv")
```


###### Stacked box plot

```{r}
svr_fit_new_diff <- svr_fit_new[svr_fit_new[,1] > 0.5,]
svr_fit_new_diff <- svr_fit_new_diff[order(svr_fit_new_diff[,1], decreasing = F),]
svr_fit_new_krt <- svr_fit_new[svr_fit_new[,2] > 0.5,]
svr_fit_new_krt <- svr_fit_new_krt[order(svr_fit_new_krt[,2], decreasing = F),]
svr_fit_new_emt <- svr_fit_new[svr_fit_new[,3] > 0.5,]
svr_fit_new_emt <- svr_fit_new_emt[order(svr_fit_new_emt[,3], decreasing = F),]
svr_fit_new_cc <- svr_fit_new[svr_fit_new[,4] > 0.5,]
svr_fit_new_cc <- svr_fit_new_cc[order(svr_fit_new_cc[,4], decreasing = F),]
svr_fit_new_others <- svr_fit_new[!rownames(svr_fit_new) %in% c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc)),]

plot.data <- data.frame(sample = rep(rownames(svr_fit_new2), 5),
                        state = rep(c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), each = nrow(svr_fit_new2)),
                        score = c(svr_fit_new2[,1],
                                  svr_fit_new2[,2],
                                  svr_fit_new2[,3],
                                  svr_fit_new2[,4],
                                  svr_fit_new2[,5]))
plot.data$sample <- factor(plot.data$sample, levels = c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                        rownames(svr_fit_new_others)))
plot.data$state <- factor(plot.data$state, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"))
# differentiated #A16BA3
# F6A000 KRT7
# FFDE00 EMT
# CC #4DD9FF
#F3A1B6 Ciliated
```

```{r, fig.width=10, fig.height=3}
ggplot(plot.data) + geom_bar(aes(y = score, x = sample, fill = state),
                           stat="identity") + theme_pubclean() + 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6"), 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  xlab("TCGA bulk tumour samples") + ylab("Proportion") +
  theme(axis.ticks.x  = element_blank(), axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size= 15),
        legend.text = element_text(size = 12),
        legend.title  = element_blank())

# ggsave("plots/Fig4_TCGA_stackedBarplot_20190509.png", width = 10, height = 5)
```

```{r}
plot.data2 <- data.frame(State = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"),
                         Fraction = colSums(svr_fit_new[,1:5] > 0)/nrow(svr_fit_new))
plot.data2$State <- factor(plot.data2$State, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")[5:1])
ggplot(plot.data2) + geom_bar(aes(y = Fraction,  x = State , fill = State),stat = "identity")+ theme_pubclean() + coord_flip()+ 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6")[5:1], 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  theme(legend.position = "none", axis.text = element_text(size = 16), axis.title = element_text(size = 20))
# ggsave("plots/Fig4_TCGA_barplot_20190509.png", width = 4, height = 5)
```


###### TSNE plot

```{r deconvolution-TCGA-PCA-plot}
colnames(svr_fit_new)[c(1,2,4,5)] <- c("Differentiated","KRT17","Cell_cycle","Ciliated")
```

```{r}
set.seed(158163)
tsne_res <- tsne::tsne(svr_fit_new[,1:5], k = 2, perplexity = round(0.05*nrow(svr_fit_new)))

df_tsne <- data.frame(TSNE_1 = tsne_res[,1],
                     TSNE_2 = tsne_res[,2],
                     EMT = svr_fit_new[,"EMT"],
                     Cell_cycle = svr_fit_new[,"Cell_cycle"],
                     Differentiated = svr_fit_new[,1],
                     KRT17 = svr_fit_new[,2],
                     Ciliated =  svr_fit_new[,5],
                     sample = rownames(svr_fit_new))
p1 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = EMT)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p2 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Cell_cycle)) + geom_point(size =1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p3 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Differentiated)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p4 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = KRT17)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
p5 <- ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Ciliated)) + geom_point(size = 1) + theme_pubclean() + scale_color_viridis() + theme(axis.title = element_blank() , legend.text = element_text(size = 7))
cowplot::plot_grid(p1,p2,p3,p4,p5, ncol = 5)
# ggsave("plots/Fig6_TSNE_deconvolution_TCGA.pdf",device = "pdf", width = 12.5, height = 3)
```

```{r}
df_tsne$Type <- "Mixture"
df_tsne$Type[df_tsne$EMT > 0.5] <- "EMT"
df_tsne$Type[df_tsne$Cell_cycle > 0.5] <- "Cell cycle"
df_tsne$Type[df_tsne$Differentiated > 0.5] <- "Differentiated"
df_tsne$Type[df_tsne$KRT17 > 0.5] <- "KRT17 Cluster"
df_tsne$Type[df_tsne$Ciliated > 0.5] <- "Ciliated"
# RColorBrewer::display.brewer.all()
RColorBrewer::brewer.pal(8, "Set1")
# "#E41A1C" "#377EB8" "#4DAF4A" "#984EA3" "#FF7F00" "#FFFF33" "#A65628" "#F781BF"
my_colour <- RColorBrewer::brewer.pal(8, "Set1")[c(2,4,6,7)]
# ggplot(df_tsne, aes(x = TSNE_1, y = TSNE_2, col = Type)) + geom_point() + theme_classic() + scale_color_manual(values=c(my_colour,"grey"))
# ggsave("plots_explore/20190423PCA_deconvolution/TSNE_deconvolution_TCGA2.png", width = 5, height = 3.5)
```


```{r}
plot.htmp <- pheatmap::pheatmap(svr_fit_new[,1:5], cluster_cols = F, show_rownames = F, border_color = NA, cellwidth = 20, cellheight = 1)
# gheat <- ggplotify::as.ggplot(plot.htmp )
# ggsave(gheat, filename = "plots/Fig4_TCGA_heatmap.pdf", device = "pdf", width = 8, height = 10, units = "in")

```

```{r survival-analysis-univariant-tcga}
# eset$emt_new <- scale(svr_fit_new[,"EMT"], scale = T, center = T)
eset$emt_new <- svr_fit_new[,"EMT"]
fit <- coxph(Surv(OS.time, OS)~emt_new, data=pData(eset))
fit
```

```{r}
summary(fit)
```

The following coding tested which cutoff of dichotomisation gives the lowest p value.

```{r}
surv_rls <- matrix(0, nrow = 1, ncol = 3)
for (i in seq(0.08, 0.57, 0.01)) {
  eset$emt_group <- NA
  eset$emt_group[eset$emt_new <= i] <- "low"
  eset$emt_group[eset$emt_new > i] <- "high"
  fit <- coxph(Surv(OS.time, OS)~emt_group, data=pData(eset))
  surv_rls <- rbind(surv_rls, c(i,summary(fit)$coef[,2], summary(fit)$coef[,5]))
}
surv_rls <- surv_rls[-1,]
```


```{r}
plot(surv_rls[,1],surv_rls[,3])
abline(h = 0.05)
```

```{r}
plot(surv_rls[,1],surv_rls[,2])
```


#### EMT & Stages

Stages split into two groups: early (I & II) and late (III & IV)

```{r  }
# restrict to only late-stage patient
eset$stage_dichotomized <- "early"
eset$stage_dichotomized[eset$clinical_stage%in% c("Stage IIIA","Stage IIIB","Stage IIIC","Stage IV")] <- "late"

fit <- coxph(Surv(OS.time, OS)~emt_new+stage_dichotomized, data=pData(eset))
fit
```

Include age and stage into survival analysis, p value for EMT is 0.04.

```{r survival-with-Age-stage-tcga}
fit <- coxph(Surv(OS.time, OS)~emt_new+stage_dichotomized+age_at_initial_pathologic_diagnosis, data=pData(eset))
fit
```

```{r}
eset$emt_group <- NA
eset$emt_group[eset$emt_new <= median(eset$emt_new)] <- "low"
eset$emt_group[eset$emt_new > median(eset$emt_new)] <- "high"
fit1 <- survfit(Surv(OS.time, OS)~emt_group, data = pData(eset))

ggsurvplot(fit1, data = pData(eset), pval = T, risk.table = T )
# pdf("plots/Fig6_TCGA_survival_median_cutoff.pdf")
# ggsurvplot(fit1, data = pData(eset), pval = T, risk.table = T )
# dev.off()
```


```{r}
table(eset$tumor_tissue_site)
```

Most of the samples were taken from ovary - primary site, which will not affect the EMT substantively.

#### EMT & age

With age as a secondary variable, the p-value for emt is 0.026.

```{r survival-withAge-tcga}
fit <- coxph(Surv(OS.time, OS)~emt_new+age_at_initial_pathologic_diagnosis, data=pData(eset))
fit
```

EMT is slightly corelated with age.

```{r EMT & age-correaltion}
cor.test(eset$emt_new, eset$age_at_initial_pathologic_diagnosis)
```

#### EMT & residual disease

No macroscopic disease is related to lower EMT.

```{r EMT & residual disease - linear regression}
rls <- glm( emt_new ~ as.character(tumor_residual_disease) == "No Macroscopic disease", 
            data = pData(eset)[eset$tumor_residual_disease != "" & eset$clinical_stage%in% c("Stage IIIA","Stage IIIB","Stage IIIC","Stage IV"),])
summary(rls)
```


```{r}
eset$residual_disease1 <- NA
eset$residual_disease1[eset$tumor_residual_disease %in% c(">20 mm", "11-20 mm") ] <- "residual"
eset$residual_disease1[eset$tumor_residual_disease %in% c("No Macroscopic disease","1-10 mm") ] <- "no_residual"

eset$emt_group <- NA
eset$emt_group[eset$emt_new > 0.1] <- "High"
eset$emt_group[eset$emt_new < 0.1] <- "Low"

# chisq.test(table(eset$emt_group[!is.na(eset$emt_group) &!is.na(eset$residual_disease1) & eset$stage_dichotomized == "Late"], 
#            eset$residual_disease1[!is.na(eset$emt_group) &!is.na(eset$residual_disease1)& eset$stage_dichotomized == "Late"]))
```

The EMT group is correlated with residual diseases for late-stage patients (p = 0.04). Higher EMT has larger residual diseases. 


```{r survival-residual-disease-tcga}
fit <- coxph(Surv(OS.time, OS)~emt_new+residual_disease1, data=pData(eset))
fit
```

```{r survival-residual-disease + stage tcga}
fit <- coxph(Surv(OS.time, OS)~emt_new+residual_disease1+stage_dichotomized, data=pData(eset))
fit
```

```{r}
cor.test(eset$emt_new, eset@assayData$exprs["RGS16", ])
```

#### Cibersort Tothill

```{r deconvolution-tothill}
svr_fit_tot <- cibersort_my(sig_matrix = sig_matrix_new, exprs_unlogged = exprs_tot)
toteset$emt.new <- svr_fit_tot[,"EMT"]
# saveRDS(svr_fit_tot, "results/20190502Deconvolution_Tothill.rds", compress = T)
# only for high-grade and late stage patients
fit_tot <- coxph(Surv(time.to.death, OS)~emt.new, data=pData(toteset)[toteset$grade >= 2 ,])
fit_tot
```

```{r}
toteset$histological.subtype2 <- "ser"
toteset$histological.subtype2[toteset$histological.subtype != "Ser"] <- "non-ser"

toteset$histological.subtype2 <- as.factor(toteset$histological.subtype2)

coxph(Surv(time.to.death, OS)~emt.new+histological.subtype2, data=pData(toteset)[toteset$grade >= 2 ,])
```


```{r}
summary(fit_tot)
```


```{r}
aocs <- pData(toteset)[toteset$grade >= 2 ,]
aocs <- aocs[!is.na(aocs$sampleID),]
aocs <- aocs[!is.na(aocs$time.to.death),]

table(aocs$grade)
```

```{r}
table(aocs$histological.subtype)
```


```{r}
surv_rls <- matrix(0, nrow = 1, ncol = 3)
for (i in seq(0, 0.80, 0.01)) {
  toteset$emt_group <- NA
  toteset$emt_group[toteset$emt.new <= i] <- "low"
  toteset$emt_group[toteset$emt.new > i] <- "high"
  fit <- coxph(Surv(time.to.death, OS)~emt_group, data=pData(toteset))
  surv_rls <- rbind(surv_rls, c(i,summary(fit)$coef[,2], summary(fit)$coef[,5]))
}
surv_rls <- surv_rls[-1,]
```


```{r}
plot(surv_rls[,1],surv_rls[,3])
abline(h = 0.05)
```

```{r}
plot(surv_rls[,1],surv_rls[,2])
```


```{r}
summary(fit_tot)
```


```{r}
toteset$emt_group <- NA
toteset$emt_group[toteset$emt.new <= median(toteset$emt.new[toteset$grade %in% c(2,3)])] <- "low"
toteset$emt_group[toteset$emt.new > median(toteset$emt.new[toteset$grade %in% c(2,3)])] <- "high"
fit1 <- survfit(Surv(time.to.death*30, OS)~emt_group, data = pData(toteset)[toteset$grade %in% c(2,3),])

ggsurvplot(fit1, data = pData(toteset)[toteset$grade %in% c(2,3),], pval = T, risk.table = T )
# pdf("plots/SuppFig6_AOCS_survival_median_cutoff.pdf")
# ggsurvplot(fit1, data = pData(toteset)[toteset$grade %in% c(2,3),], pval = T, risk.table = T )
# dev.off()
```

```{r}
toteset$emt_group <- NA
toteset$emt_group[toteset$emt.new <= quantile(toteset$emt.new, 0.3)] <- "low"
toteset$emt_group[toteset$emt.new >= quantile(toteset$emt.new, 0.7)] <- "high"

fit_tot <- coxph(Surv(time.to.death, OS)~emt_group, data=pData(toteset)[toteset$grade >= 2 ,])
fit_tot

```

```{r}
quantile(toteset$emt.new, 0.7)
```

```{r}
plot.htmp <- pheatmap::pheatmap(svr_fit_tot[,1:5], cluster_cols = F, show_rownames = F, border_color = NA, cellwidth = 20, cellheight = 1)
gheat <- ggplotify::as.ggplot(plot.htmp )
# ggsave(gheat, filename = "plots/Fig4_Tothill_heatmap.pdf", device = "pdf", width = 8, height = 10, units = "in")
```

```{r survival-age-tothill}
fit_tot <- coxph(Surv(time.to.death, OS)~emt.new+age, data=pData(toteset))
fit_tot
```


###### Stacked barplot
```{r}
svr_fit_new_diff <- svr_fit_tot[svr_fit_tot[,1] > 0.5,]
svr_fit_new_diff <- svr_fit_new_diff[order(svr_fit_new_diff[,1], decreasing = F),]
svr_fit_new_krt <- svr_fit_tot[svr_fit_tot[,2] > 0.5,]
svr_fit_new_krt <- svr_fit_new_krt[order(svr_fit_new_krt[,2], decreasing = F),]
svr_fit_new_emt <- svr_fit_tot[svr_fit_tot[,3] > 0.5,]
svr_fit_new_emt <- svr_fit_new_emt[order(svr_fit_new_emt[,3], decreasing = F),]
svr_fit_new_cc <- svr_fit_tot[svr_fit_tot[,4] > 0.5,]
svr_fit_new_cc <- svr_fit_new_cc[order(svr_fit_new_cc[,4], decreasing = F),]
svr_fit_new_cil <- svr_fit_tot[svr_fit_tot[,5] > 0.5,]
svr_fit_new_cil <- svr_fit_new_cil[order(svr_fit_new_cil[,5], decreasing = F),]
svr_fit_new_others <- svr_fit_tot[!rownames(svr_fit_tot) %in% c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                               rownames(svr_fit_new_cil)),]

plot.data <- data.frame(sample = rep(rownames(svr_fit_tot), 5),
                        state = rep(c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), each = nrow(svr_fit_tot)),
                        score = c(svr_fit_tot[,1],
                                  svr_fit_tot[,2],
                                  svr_fit_tot[,3],
                                  svr_fit_tot[,4],
                                  svr_fit_tot[,5]))
plot.data$sample <- factor(plot.data$sample, levels = c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                        rownames(svr_fit_new_cil),
                                                        rownames(svr_fit_new_others)))
plot.data$state <- factor(plot.data$state, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"))
# differentiated #A16BA3
# F6A000 KRT7
# FFDE00 EMT
# CC #4DD9FF
#F3A1B6 Ciliated
```

```{r, fig.width=10, fig.height=3}
ggplot(plot.data) + geom_bar(aes(y = score, x = sample, fill = state),
                           stat="identity") + theme_pubclean() + 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6"), 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  xlab("AOCS bulk tumour samples") + ylab("Proportion") +
  theme(axis.ticks.x  = element_blank(), axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size= 15),
        legend.text = element_text(size = 12),
        legend.title  = element_blank())

# ggsave("plots/Fig4_AOCS_stackedBarplot_20190509.png", width = 10, height = 5)
```

```{r}
plot.data2 <- data.frame(State = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"),
                         Fraction = colSums(svr_fit_tot[,1:5] > 0)/nrow(svr_fit_tot))
plot.data2$State <- factor(plot.data2$State, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")[5:1])
ggplot(plot.data2) + geom_bar(aes(y = Fraction,  x = State , fill = State),stat = "identity")+ theme_pubclean() + coord_flip()+ 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6")[5:1], 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  theme(legend.position = "none", axis.text = element_text(size = 16), axis.title = element_text(size = 20))
# ggsave("plots/Fig4_AOCS_barplot_20190509.png", width = 4, height = 5)
```


## Heatmap: TCGA&AOCS

```{r}
data.heatmap <- t(svr_fit_new[,1:5])
pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
  show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)

# pdf("plots/Fig4_heatmap_mini_tcga_20190510.pdf", height = 5, width = 5)
# pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
#   show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)
# dev.off()
```

```{r}
data.heatmap <- t(svr_fit_tot[,1:5])
pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
  show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)

# pdf("plots/Fig4_heatmap_mini_aocs_20190510.pdf", height = 5, width = 5)
# pheatmap::pheatmap(data.heatmap, color = colorRampPalette(brewer.pal(n = 7, name = "Blues"))(100),
#   show_colnames = F, show_rownames = T, cluster_rows = F, cellheight = 30, cellwidth = 0.3)
# dev.off()
```

#### Cibersort OXO-PCR

```{r cibersort oxopcr}
svr_fit_oxo <- cibersort_my(sig_matrix = sig_matrix_new, exprs_unlogged = cpm(as.matrix(oxopcr_pre)))
```

```{r oxopcr_heatmap}
rownames(svr_fit_oxo) <- gsub(rownames(svr_fit_oxo), pattern = "X", replacement = "")
colnames(svr_fit_oxo)[1:5] <- c("Differentiated", "KRT17","EMT","Cell cycle","Ciliated")
rownames(svr_fit_oxo)[rownames(svr_fit_oxo) == "1G3_new"] <- "1016_1G3new"
```


###### Heatmap

```{r}
plot.htmp <- pheatmap::pheatmap(svr_fit_oxo[order(oxopcr_info_pre$`PATIENT ID`),1:5], cluster_cols = F, cluster_rows = F,show_rownames = T, border_color = NA, cellwidth = 20, cellheight = 8, fontsize = 8)
gheat <- ggplotify::as.ggplot(plot.htmp )
# ggsave(gheat, filename = "plots/Fig4_OXOPCR_heatmap20190316.pdf", device = "pdf", width = 8, height = 12, units = "in")
```

###### Stacked barplot

```{r}
svr_fit_new_diff <- svr_fit_oxo[svr_fit_oxo[,1] > 0.5,]
svr_fit_new_diff <- svr_fit_new_diff[order(svr_fit_new_diff[,1], decreasing = F),]
svr_fit_new_krt <- svr_fit_oxo[svr_fit_oxo[,2] > 0.5,]
svr_fit_new_krt <- svr_fit_new_krt[order(svr_fit_new_krt[,2], decreasing = F),]
svr_fit_new_emt <- svr_fit_oxo[svr_fit_oxo[,3] > 0.5,]
svr_fit_new_emt <- svr_fit_new_emt[order(svr_fit_new_emt[,3], decreasing = F),]
svr_fit_new_cc <- svr_fit_oxo[svr_fit_oxo[,4] > 0.5,]
svr_fit_new_cc <- svr_fit_new_cc[order(svr_fit_new_cc[,4], decreasing = F),]
svr_fit_new_others <- svr_fit_oxo[!rownames(svr_fit_oxo) %in% c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc)),]

plot.data <- data.frame(sample = rep(rownames(svr_fit_oxo), 5),
                        state = rep(c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                                    each = nrow(svr_fit_oxo)),
                        score = c(svr_fit_oxo[,1],
                                  svr_fit_oxo[,2],
                                  svr_fit_oxo[,3],
                                  svr_fit_oxo[,4],
                                  svr_fit_oxo[,5]))
plot.data$sample <- factor(plot.data$sample, levels = c(rownames(svr_fit_new_diff),rownames(svr_fit_new_krt),
                                                               rownames(svr_fit_new_emt),rownames(svr_fit_new_cc),
                                                        rownames(svr_fit_new_others)))
plot.data$state <- factor(plot.data$state, levels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"))
# differentiated #A16BA3
# F6A000 KRT7
# FFDE00 EMT
# CC #4DD9FF
#F3A1B6 Ciliated
```

```{r, fig.width=10, fig.height=3}
ggplot(plot.data) + geom_bar(aes(y = score, x = sample, fill = state),
                           stat="identity") + theme_pubclean() + 
  scale_fill_manual(values = c("#A16BA3","#F6A000","#FFDE00","#4DD9FF","#F3A1B6"), 
                    breaks = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated"), 
                    labels = c("Differentiated","KRT17","EMT","Cell cycle","Ciliated")) +
  xlab("TCGA bulk tumour samples") + ylab("Proportion") +
  theme(axis.ticks.x  = element_blank(), axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title = element_text(size= 15),
        legend.text = element_text(size = 12),
        legend.title  = element_blank())

# ggsave("plots/Fig5_OXOPCR_stackedBarplot_20190509.png", width = 7, height = 3)
```

###### Violin plots

```{r}
oxopcr_info_pre <- cbind(oxopcr_info_pre, svr_fit_oxo)
rownames(oxopcr_info_pre) <- oxopcr_info_pre$sample.id
# write.csv(oxopcr_info_pre[order(oxopcr_info_pre$`PATIENT ID`),], "results/20190316_OXOPCR_cibersort_output.csv")

oxosce <- SingleCellExperiment(assays = list(counts = as.matrix(oxopcr_pre)), colData = oxopcr_info_pre)
logcounts(oxosce) <- log1p(calculateCPM(oxosce))
oxosce <- normalize(oxosce)
oxosce$patient <- as.factor(as.character(oxosce$`PATIENT ID`))

oxosce$EMT_group <- "EMT-middle T"
oxosce$EMT_group[oxosce$EMT >= quantile(oxosce$EMT, 2/3)] <- "EMT-high T"
oxosce$EMT_group[oxosce$EMT <= quantile(oxosce$EMT, 1/3)] <- "EMT-low T"

# oxosce$EMT_group[oxosce$EMT >= quantile(oxosce$EMT[oxosce$type != "stroma"], 2/3)] <- "EMT-high T"
# oxosce$EMT_group[oxosce$EMT <= quantile(oxosce$EMT[oxosce$type != "stroma"], 1/3)] <- "EMT-low T"
# oxosce$EMT_group[oxosce$type == "stroma"] <- "Stroma"
```

```{r}
plotExpression(oxosce, 
               features = c("EPCAM","KRT7","MUC16","WT1"), 
               x = "EMT_group", ncol = 2, scales ="free", colour_by = "patient") + xlab("") +
    theme(strip.text.x = element_text(size=12, angle=0, face = "italic"),
        strip.background = element_rect(colour="black", fill="white"))

# ggsave("plots/SuppFig4_violinplots_OXO-pcr_tumor_markers.png", width = 7,height = 5)
```



## Permutation

For TCGA data

```{r permutation-tcga}
Perm <- 500
prob <- 0.9
# ncol(eset) * 0.8
#  246.4 
p_val_tcga <- c()

for(itor in 1:Perm){
  idx <- sample(x = 1:ncol(eset), size = round(ncol(eset) * prob), replace = F)
  fit <- coxph(Surv(OS.time, OS)~emt_new, data=pData(eset)[idx,])
  p_val_tcga <- c(p_val_tcga, summary(fit)$coef[,5])
}

sum( p_val_tcga <= 0.05)/500 * 100
# [1] 95.2
```

For the high-grade tumours in Tothill's data

```{r}
Perm <- 500
prob <- 0.9
# ncol(eset) * 0.8
#  246.4 
p_val_tot <- c()

for(itor in 1:Perm){
  idx <- sample(x = which(toteset$Grade >= 2), size = round(sum(toteset$Grade >= 2, na.rm = TRUE) * prob), replace = F)
  fit <- coxph(Surv(time.to.death, OS)~emt.new, data=pData(toteset)[idx,])
  p_val_tot <- c(p_val_tot, summary(fit)$coef[,5])
}

sum( p_val_tot <= 0.05)/500 * 100
# [1] 99.4
```



## Power calculation

http://www.evolutionarystatistics.org/document.pdf

```{r}
library(powerSurvEpi)
hist(eset$emt_new)
```

```{r}
eset$emt_strata <- NA
eset$emt_strata[eset$emt_new >= quantile(eset$emt_new, 2/3)] <- "High"
eset$emt_strata[eset$emt_new <= quantile(eset$emt_new, 1/3)] <- "Low"

eset$emt_strata <- factor(eset$emt_strata, levels = c("Low","High"))
eset$stage_dichotomized <- "Early"
eset$stage_dichotomized[eset$clinical_stage %in% c("Stage IIIA","Stage IIIB","Stage IIIC","Stage IV")] <- "Late"
fit <- coxph(Surv(OS.time, OS)~emt_strata+stage_dichotomized, data=pData(eset))
fit
```


```{r}
table(eset$emt_strata)
```

```{r}
dim(eset)
```


```{r}
sfit <- survfit(Surv(OS.time, OS)~emt_strata, data=pData(eset))
ggsurvplot(sfit, pval = T, legend= "right") 
```

```{r}
df_pct <- data.frame(OS.time = eset$OS.time,
                     OS = eset$OS,
                     emt = NA)
df_pct$emt[eset$emt_strata == "Low"] <- "C"
df_pct$emt[eset$emt_strata == "High"] <- "E"
df_pct <- na.omit(df_pct)
df_pct$emt <- as.factor(df_pct$emt)
```

```{r}
PCT150 <- powerCT(Surv(OS.time, OS)~emt, df_pct, nE=150,
                  nC=150, RR=1.65)
print(PCT150$power)
```

```{r}
ss <- ssizeCT(Surv(OS.time, OS)~emt, df_pct,
              power=0.8, RR=1.65, k=1)
print(ss$ssize)
```

## CuratedOvarianCancer database

```{r}
# source("codings/20181123CIBERSORT_sigMatrix_function.R")
esets <- readRDS("rds/20190213CuratedOvarianData_esets.rds")
sig_matrix_new <- readRDS("rds/20190213sig_matrix_new.rds")

svr_fit <- list()
fit <- list()
esets_tumor <- list()

for(itor in 1:length(esets)) { # go through all the
 
  ## tumor data
  eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
  
  ## deconvolution
  if(max(na.omit(eset_tmp@assayData$exprs)) < 50) { ## log-space
    svr_fit[[itor]] <- cibersort_my(sig_matrix = sig_matrix_new, 
                                    exprs_unlogged = exp(na.omit(eset_tmp@assayData$exprs)))
  } else {         # non-log space
    svr_fit[[itor]] <- cibersort_my(sig_matrix = sig_matrix_new, 
                                    exprs_unlogged = eset_tmp@assayData$exprs)
  }
}

for(itor in 1:length(esets)) { # go through all the
 
  ## tumor data
  eset_tmp <- esets[[itor]][,esets[[itor]]$sample_type %in% c("tumor","metastatic")]
  
  # # survival data
  eset_tmp$y <- Surv(eset_tmp$days_to_death,
                     eset_tmp$vital_status == "deceased")
  eset_tmp$emt <- svr_fit[[itor]][,"EMT"]

  # ## serous subtype
  # if(any(!is.na(eset_tmp$histological_type))) {
  #   eset_tmp <- eset_tmp[,eset_tmp$histological_type == "ser"]
  # }
  # 
  # ## grades
  # if(any(!is.na(eset_tmp$grade))) {
  #   eset_tmp <- eset_tmp[,eset_tmp$grade %in% c(2,3)]
  # }
  # 
  # survival test
  # fit[[itor]] <- coxph(eset_tmp$y~eset_tmp$emt) 
  eset_tmp$histological_type2 <- NA
  eset_tmp$histological_type2[eset_tmp$histological_type=="ser"] <- "ser"
  eset_tmp$histological_type2[eset_tmp$histological_type!="ser"] <- "non-ser"
  eset_tmp$histological_type2[is.na(eset_tmp$histological_type)] <- NA
  
  eset_tmp$grade2 <- NA
  eset_tmp$grade2[eset_tmp$grade==1] <- "low"
  eset_tmp$grade2[eset_tmp$grade>1] <- "high"
  eset_tmp$grade2[is.na(eset_tmp$grade)] <- NA
  
  if(all(is.na(eset_tmp$grade))) {
    eset_tmp$grade2 <- "high"
  }
  
  if(any(eset_tmp$grade2 %in% "low") & any(eset_tmp$grade2 %in% "non-ser")){
    fit[[itor]] <- coxph(y~emt+grade2+histological_type2, eset_tmp ) 
    eset_tmp <- eset_tmp[,!is.na(eset_tmp$grade2) & !is.na(eset_tmp$histological_type)]
  } else if (any(eset_tmp$grade2 %in% "low") & !any(eset_tmp$grade2 %in% "non-ser")){
    fit[[itor]] <- coxph(y~emt+grade2, eset_tmp ) 
    eset_tmp <- eset_tmp[,!is.na(eset_tmp$grade2) ]
  } else if (!any(eset_tmp$grade2 %in% "low") & any(eset_tmp$grade2 %in% "non-ser")){
    fit[[itor]] <- coxph(y~emt+histological_type2, eset_tmp ) 
    eset_tmp <- eset_tmp[,!is.na(eset_tmp$histological_type)]
  } else {
    fit[[itor]] <- coxph(y~emt, eset_tmp ) 
  }
  # fit2 <- coxph(eset_tmp$y~eset_tmp$emt+eset_tmp$grade2) 
  # fit2
  esets_tumor[[itor]] <- eset_tmp
}

df <- matrix(NA, nrow = 7, ncol = 6)
colnames(df) <- c("nsamples","nevent","hazard_ratio","pvalue","lower.CI95","upper.CI95")
df <- as.data.frame(df)
for(itor in 1:7){
  df$lower.CI95[itor] <- summary(fit[[itor]])$conf.int[1,3]
  df$upper.CI95[itor] <- summary(fit[[itor]])$conf.int[1,4]
  df$hazard_ratio[itor] <- summary(fit[[itor]])$conf.int[1,1]
  df$pvalue[itor] <- summary(fit[[itor]])$coefficient[1,5]
  df$nsamples[itor] <- summary(fit[[itor]])$n
  df$nevent[itor] <- summary(fit[[itor]])$nevent
}

df$database <- c("E.MTAB.386", "GSE13876", "GSE26193",
              "GSE26712", "GSE32062.GPL6480",
              "GSE49997", "GSE51088")
  
# write.csv(df, "results/20190503CuratedOvarianData_results_multivariable_test.csv", row.names = F)

## IC95 error bar plot----
df <- df[order(df$hazard_ratio, decreasing = T),]
df <- df[!grepl("TCGA",df$database),]
ggplot(df[which(df$nsamples > 100),], aes(y = database, x = hazard_ratio))+ geom_point(shape=21, fill = "black",size=4) +
  geom_errorbarh(aes(xmax = upper.CI95, xmin = lower.CI95), height=0.2, size=1) + theme_survminer() +xlim(c(0, 10)) +
  geom_vline(xintercept = 1, lty = 3, col = "black") + xlab("Hazard ratio") +ylab("")

# ggsave("plots_explore/20190213Hazard_ratio_CI95_CuratedOvarianData.png",width = 5, height = 6)

```

###### Put all datasets together

```{r}
toteset_hg <- toteset[,toteset$grade %in% c(2,3)]
```


```{r}
## put all datasets together

df_all <- data.frame(database = c(rep("E.MTAB.386",ncol(esets_tumor[[1]])),
                                  rep("GSE13876",ncol(esets_tumor[[2]])),
                                  rep("GSE26193",ncol(esets_tumor[[3]])),
                                  rep("GSE26712",ncol(esets_tumor[[4]])),
                                  rep("GSE32062.GPL6480",ncol(esets_tumor[[5]])),
                                  rep("GSE49997",ncol(esets_tumor[[6]])),
                                  rep("GSE51088",ncol(esets_tumor[[7]])),
                                  rep("AOCS", ncol(toteset_hg)),
                                  rep("TCGA", ncol(eset))),
                     
                     sample_type = c(esets_tumor[[1]]$sample_type,esets_tumor[[2]]$sample_type,
                                     esets_tumor[[3]]$sample_type,esets_tumor[[4]]$sample_type,
                                     esets_tumor[[5]]$sample_type,esets_tumor[[6]]$sample_type,
                                     esets_tumor[[7]]$sample_type, rep("tumor",ncol(toteset_hg)),
                                     rep("tumor", ncol(eset))),
                     grade = c(rep(2, ncol(esets_tumor[[1]])),esets_tumor[[2]]$grade,
                               esets_tumor[[3]]$grade,
                               rep(2,ncol(esets_tumor[[4]])),esets_tumor[[5]]$grade,
                               esets_tumor[[6]]$grade,esets_tumor[[7]]$grade, 
                               toteset_hg$grade, 
                               rep(2, ncol(eset))),
                     histological_type = c(esets_tumor[[1]]$histological_type,esets_tumor[[2]]$histological_type,
                                           esets_tumor[[3]]$histological_type,esets_tumor[[4]]$histological_type,
                                        esets_tumor[[5]]$histological_type,esets_tumor[[6]]$histological_type,
                                        esets_tumor[[7]]$histological_type, tolower(toteset_hg$histological.subtype),
                                        rep("ser", ncol(eset)))
                     )

toteset_hg$vital_status <- NA
toteset_hg$vital_status[toteset_hg$patient.status != "D"] <- "living"
toteset_hg$vital_status[toteset_hg$patient.status == "D"] <- "deceased"

df_all$OS <- c(esets_tumor[[1]]$vital_status,esets_tumor[[2]]$vital_status,
               esets_tumor[[3]]$vital_status,esets_tumor[[4]]$vital_status,
               esets_tumor[[5]]$vital_status,esets_tumor[[6]]$vital_status,
               esets_tumor[[7]]$vital_status,toteset_hg$vital_status, tolower(eset$vital_status)
               )

df_all$days_to_death <- c(esets_tumor[[1]]$days_to_death,esets_tumor[[2]]$days_to_death,
                          esets_tumor[[3]]$days_to_death,esets_tumor[[4]]$days_to_death,
                          esets_tumor[[5]]$days_to_death,esets_tumor[[6]]$days_to_death,
                          esets_tumor[[7]]$days_to_death,
                          toteset_hg$time.to.death*30,
                          eset$OS.time)

df_all <- df_all[df_all$sample_type %in% c("metastatic","tumor"),]

df_all$EMT <- c( esets_tumor[[1]]$emt, esets_tumor[[2]]$emt,
                  esets_tumor[[3]]$emt,  esets_tumor[[4]]$emt,
                  esets_tumor[[5]]$emt, esets_tumor[[6]]$emt,
                  esets_tumor[[7]]$emt,
                 toteset_hg$emt.new, eset$EMT)

df_all$grade2 <- NA
df_all$grade2[df_all$grade == 1] <- "Low-grade"
df_all$grade2[df_all$grade > 1 ] <- "High-grade"

df_all$histological_type2 <- NA
df_all$histological_type2[df_all$histological_type == "ser"] <- "serous"
df_all$histological_type2[df_all$histological_type != "ser"] <- "non-serous"

# saveRDS(fit,"results/20181122CuratedOvarianData_survival_list.rds",compress = T)
df_all$years_to_death <- df_all$days_to_death/365
df_all$y <- Surv(df_all$years_to_death,
                     df_all$OS == "deceased")

fit <- coxph(y~EMT+grade2+histological_type2, data = df_all) 
summary(fit)
```



```{r}
ggplot(df_all, aes(x = database, y = EMT)) + geom_violin(scale = "width") + theme_classic() + geom_jitter(alpha = 0.3, aes(col = database))
# ggsave("plots_explore/EMT_curatedOvarianData.png", height = 4, width = 10)
```

```{r}
ggplot(df_all, aes(x = database, col = OS, y = days_to_death)) + geom_violin(scale = "width") + theme_classic() + geom_jitter(alpha = 0.3, aes(col = OS))
# ggsave("plots_explore/curatedOvarianData_diff_censorDuration.png", height = 4, width = 10)
```

###### Overall KM plot

```{r}
table(df_all$database[!is.na(df_all$years_to_death)])
```

```{r}
df
```


```{r}
df_all$EMT_group <- NA
# df_all$EMT_group[df_all$EMT >= median(df_all$EMT)] <- "high"

for(itor in 1:9){
   emt_tmp <- median(df_all$EMT[df_all$database == unique(df_all$database)[itor]])
   df_all$EMT_group[df_all$database == unique(df_all$database)[itor] & df_all$EMT > emt_tmp] <- "high"
   df_all$EMT_group[df_all$database == unique(df_all$database)[itor] & df_all$EMT <= emt_tmp] <- "low"
}

table(df_all$EMT_group, df_all$database)

```

```{r}
fit1 <- survfit(y~EMT_group, data = df_all)
ggsurvplot(fit1, data = df_all, pval = F) + xlab("Years") 

# ggsave("plots/20190503_survival_pooled_nine.png", width = 5, height = 4)
```



```{r}
fit <- coxph(y~EMT_group, data = df_all) 
summary(fit)
```

## Add NCBI Entrez gene ID

```{r}
## Add NCBI gene ID
list <- read.csv("../../../../Patent - OvCa prognosis/Signatures_list20190219.csv", as.is = T, row.names = 1)
library(biomaRt)
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
filters = listAttributes(ensembl)
filters[1:5,]

dat = getBM(filters = "hgnc_symbol",attributes = c( "hgnc_symbol","entrezgene","ensembl_gene_id"),
            values = list$Gene.Symbol, mart = ensembl)

list$Ensembl_gene_id <- NA
list$Entrez_gene_id <- NA
for(itor in 1:nrow(list)){
  tmp <- unique(dat$ensembl_gene_id[dat$hgnc_symbol == list$Gene.Symbol[itor]])
  if(length(tmp) > 1) {
    list$Ensembl_gene_id[itor] <- paste(tmp, collapse = " ,")
  } else if (length(tmp)== 1) {
    list$Ensembl_gene_id[itor] <- tmp
  }
  
   tmp <- unique(dat$entrezgene[dat$hgnc_symbol == list$Gene.Symbol[itor]])
  if(length(tmp) > 1) {
    list$Entrez_gene_id[itor] <- paste(tmp, collapse = ",")
  } else if (length(tmp)== 1) {
    list$Entrez_gene_id[itor] <- tmp
  }
}

# write.csv(list, "../../../../Patent - OvCa prognosis/Signatures_list20190219_updated.csv",  row.names = T)
```

## Characterise the EMT-high in TCGA

#### Markers of EMT-high

```{r}
dge <- DGEList(counts=eset@assayData$exprs[,eset$emt_strata %in% c("Low","High")])
A <- rowSums(dge$counts) # filter genes
dge <- dge[A > 10,, keep.lib.sizes=FALSE]
dge <- calcNormFactors(dge)

group <- eset$emt_strata[ eset$emt_strata %in% c("Low","High")]
design <- model.matrix(~0+group) # model by EMT groups

# limma
v <- voom(dge, design, plot=TRUE)
fit <- lmFit(v, design)

cont.matrix <- makeContrasts(groupHigh-groupLow,
                             levels = design)
fit <- contrasts.fit(fit, cont.matrix)
fit <- eBayes(fit)

marker_EMThigh <- topTable(fit,number = Inf, p.value = 0.05, lfc = 1, coef = 1)
marker_EMThigh$gene <- rownames(marker_EMThigh)
marker_EMThigh <- marker_EMThigh[order(marker_EMThigh$logFC, decreasing = T),]
head(marker_EMThigh)

# write.csv(marker_EMThigh[marker_EMThigh$logFC >= 1,],"results/TableS7_markers_EMThigh20190215.csv")

```


```{r}
df_plot <- data.frame(exprs = c(eset@assayData$exprs["TWIST1",],eset@assayData$exprs["TWIST2",],eset@assayData$exprs["SNAI2",]), 
                      group = rep(eset$emt_strata,3), 
                      gene = rep(c("TWIST1","TWIST2","SNAI2"), each = ncol(eset)))
df_plot <- na.omit(df_plot)
df_plot$exprs <- log1p(df_plot$exprs)
df_plot$group <- paste("EMT-", df_plot$group, sep = "")


ggplot(df_plot, aes(x = group, y = exprs)) + 
  geom_violin() + theme_pubr() + 
  geom_jitter(shape=16, alpha = 0.6, position=position_jitter(0.2)) + 
  xlab("") + ylab("Expression level") +
  facet_grid(~gene,margins = "am", scales = "free") +
  theme(strip.text.x = element_text(size=12, angle=0, face = "italic"),
        strip.background = element_rect(colour="black", fill="white"))

# ggsave("plots/Fig4_TWIST_SNAI2_violin20190215.png", width = 6, height = 4)
# ggsave("plots/Fig4_TWIST_SNAI2_violin20190215.pdf", width = 6, height = 3)
```


#### miRNA

miRNA mature strand expression RNAseq

```{r}
mirna <- read.delim("../TCGA/Xena/TCGA/miRNA/OV-TP.membership.txt", as.is = T)
eset$miRNA_cluster <- mirna$membership.1[match(colnames(eset), rownames(mirna))]

ggplot(pData(eset)[!is.na(eset$miRNA_cluster),], aes(x = as.factor(miRNA_cluster), y = emt_new)) + geom_violin() + theme_classic() + geom_hline(yintercept = quantile(eset$emt_new,1/3), lty = 2) + geom_hline(yintercept = quantile(eset$emt_new,2/3), lty = 2) + geom_jitter(shape=16, alpha = 0.6, position=position_jitter(0.2), aes(col = emt_group)) + xlab("miRNA clusters")

# ggsave("plots_explore/20181128TCGA_EMT/EMT_miRNA.png", width = 5, height = 3)
```

```{r}
table(eset$miRNA_cluster, eset$emt_strata)

t.test(eset$emt_new[eset$miRNA_cluster == 1], eset$emt_new[eset$miRNA_cluster > 1])
```

EMT high is enriched in the miRNA cluster 1.


###### DE analysis of miRNA

```{r}
miRNA <- read.delim("../TCGA/Xena/TCGA/miRNA/gdac.broadinstitute.org_OV.Merge_mirnaseq__illuminahiseq_mirnaseq__bcgsc_ca__Level_3__miR_gene_expression__data.Level_3.2016012800.0.0/OV.mirnaseq__illuminahiseq_mirnaseq__bcgsc_ca__Level_3__miR_gene_expression__data.data.txt", sep = "\t",row.names = 1,as.is = T)
# rownames(miRNA) <- miRNA[,1]
miRNA <- miRNA[,miRNA[1,] == "read_count"]
colnames(miRNA) <- gsub(pattern = "[.]", replacement = "-", x = colnames(miRNA))
colnames(miRNA) <- substr(colnames(miRNA), start = 1, stop = 15)

miRNA <- miRNA[,na.omit(match(colnames(eset), colnames(miRNA)))]
miRNA <- miRNA[-1,]

for(itor in 1:ncol(miRNA)){ ## transfer it to numeric
  miRNA[,itor] <- as.numeric(miRNA[,itor])
}

group <- eset$emt_strata[match(colnames(miRNA), colnames(eset))]

dge <- DGEList(counts=miRNA[,!is.na(group)])
A <- rowSums(dge$counts) # filter genes
dge <- dge[A > 10,, keep.lib.sizes=FALSE]
dge <- calcNormFactors(dge)

group <- na.omit(group)
design <- model.matrix(~0+group) # model by EMT groups

# limma
v <- voom(dge, design, plot=TRUE)
fit <- lmFit(v, design)

cont.matrix <- makeContrasts(groupHigh-groupLow, levels = design)
fit <- contrasts.fit(fit, cont.matrix)
fit <- eBayes(fit)

de_genes <- topTable(fit, number = Inf, p.value = 0.05)
de_genes$gene <- rownames(de_genes)
de_genes <- de_genes[order(de_genes$logFC, decreasing = T),]
# write.csv(de_genes, file = "results/miRNA_TCGA/20181211miRNA_DE_EMTHighvsLow.csv")

head(de_genes)
```


The top DE miRNAs are reported here: https://link.springer.com/article/10.1007/s11914-014-0229-9

miRNA-483 inhibites ECM?? https://www.ncbi.nlm.nih.gov/pubmed/26747772

miR-200 family (miR-200a, miR-200b, miR-200c, miR-141, and miR-429, which share a consensus seed sequence) have emerged as new epithelial markers and repressors of EMT: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3686549/

###### Volcano plot

```{r}
## http://bioinformatics.knowledgeblog.org/2011/06/21/volcano-plots-of-microarray-data/

tt <- topTable(fit, number = Inf)
tt$names <- gsub(rownames(tt), pattern = "hsa-mir", replacement = "miR")

##Highlight genes that have an absolute fold change > 2 and a p-value < Bonferroni cut-off
tt$threshold <- as.factor(abs(tt$logFC) > 0.5 & tt$adj.P.Val < 0.05)
tt$threshold2 <- as.factor(abs(tt$logFC) > 1.5 & tt$adj.P.Val < 0.001)
tt$threshold2[tt$names %in% c("miR-200a","miR-200b","miR-200c","miR-141","miR-429")] <- T
tt$threshold2[tt$names %in% c("miR-514-2","miR-514-3","miR-509-2","miR-509-3")] <- F
##Construct the plot object
ggplot(data=tt, aes(x=logFC, y=-log10(P.Value), colour=paste(threshold, threshold2))) +
  geom_point(alpha=0.4, size=1.75) +
  xlim(c(-3, 3)) + ylim(c(0, 13)) +
  xlab("log2 fold change") + ylab("-log10 p-value")+ 
  geom_text_repel(data = tt[tt$threshold2 == TRUE,],aes(x = logFC, y=-log10(P.Value),label=names) , colour="black", alpha = 0.75) + theme_pubr() + theme(legend.position = "none")

# ggsave("plots/Fig_miRNA_volcano_plot20190215.png", width = 5, height = 4)
```


## AOCS: Grade & ciliated cells

```{r}
toteset$ciliated.new <- svr_fit_tot[,"ciliated"]
ggplot(as.data.frame(pData(toteset))[!is.na(toteset$grade),], aes(x = as.factor(grade), y = ciliated.new)) + geom_violin(scale = "width") + theme_classic() + geom_quasirandom(alpha = 0.4, width = 0.3) +
  xlab("Grades")+ ylab("Score of ciliated signature")

# ggsave("plots/20190215AOCS_grade_ciliated.pdf", width = 4, height = 3)
```

```{r}
wilcox.test(toteset$ciliated.new[toteset$grade==1], toteset$ciliated.new[toteset$grade>1], alternative = "greater")
```

## Technical

```{r}
sessionInfo()
```

R version 3.5.2 (2018-12-20)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.4

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] bindrcpp_0.2.2              GEOquery_2.50.5             logging_0.9-107             genefilter_1.64.0           curatedOvarianData_1.20.0  
 [6] affy_1.60.0                 factoextra_1.0.5            RColorBrewer_1.1-2          edgeR_3.24.3                limma_3.38.3               
[11] viridis_0.5.1               viridisLite_0.3.0           ggrepel_0.8.0               ggbeeswarm_0.6.0            scater_1.10.1              
[16] SingleCellExperiment_1.4.1  SummarizedExperiment_1.12.0 DelayedArray_0.8.0          BiocParallel_1.16.5         matrixStats_0.54.0         
[21] GenomicRanges_1.34.0        GenomeInfoDb_1.18.1         powerSurvEpi_0.1.0          survival_2.43-3             survminer_0.4.3            
[26] ggpubr_0.2                  magrittr_1.5                ggplot2_3.1.0               preprocessCore_1.44.0       e1071_1.7-0.1              
[31] xbioc_0.1.15                AnnotationDbi_1.44.0        IRanges_2.16.0              S4Vectors_0.20.1            MASS_7.3-51.1              
[36] bseqsc_1.0                  Biobase_2.42.0              BiocGenerics_0.28.0         csSAM_1.4                   Rcpp_1.0.0                 

loaded via a namespace (and not attached):
  [1] backports_1.1.3          NMF_0.22                 plyr_1.8.4               lazyeval_0.2.1           splines_3.5.2           
  [6] gridBase_0.4-7           digest_0.6.18            foreach_1.4.4            BiocInstaller_1.32.1     memoise_1.1.0           
 [11] cluster_2.0.7-1          doParallel_1.0.14        readr_1.3.1              annotate_1.60.0          colorspace_1.4-0        
 [16] blob_1.1.1               xfun_0.4                 dplyr_0.7.8              crayon_1.3.4             RCurl_1.95-4.11         
 [21] bindr_0.1.1              zoo_1.8-4                iterators_1.0.10         glue_1.3.0               registry_0.5            
 [26] gtable_0.2.0             zlibbioc_1.28.0          XVector_0.22.0           kernlab_0.9-27           Rhdf5lib_1.4.2          
 [31] prabclus_2.2-7           DEoptimR_1.0-8           HDF5Array_1.10.1         scales_1.0.0             pheatmap_1.0.12         
 [36] mvtnorm_1.0-8            DBI_1.0.0                rngtools_1.3.1           bibtex_0.4.2             xtable_1.8-3            
 [41] cmprsk_2.2-7             gridGraphics_0.3-0       bit_1.1-14               mclust_5.4.2             km.ci_0.5-2             
 [46] Formula_1.2-3            tsne_0.1-3               fpc_2.1-11.1             modeltools_0.2-22        XML_3.98-1.17           
 [51] pkgconfig_2.0.2          flexmix_2.3-14           nnet_7.3-12              locfit_1.5-9.1           ggplotify_0.0.3         
 [56] tidyselect_0.2.5         labeling_0.3             rlang_0.3.1              reshape2_1.4.3           munsell_0.5.0           
 [61] tools_3.5.2              generics_0.0.2           RSQLite_2.1.1            broom_0.5.1              stringr_1.4.0           
 [66] yaml_2.2.0               knitr_1.21               bit64_0.9-7              robustbase_0.93-3        survMisc_0.5.5          
 [71] purrr_0.3.0              dendextend_1.9.0         nlme_3.1-137             whisker_0.3-2            xml2_1.2.0              
 [76] compiler_3.5.2           rstudioapi_0.9.0         curl_3.3                 beeswarm_0.2.3           affyio_1.52.0           
 [81] tibble_2.0.1             stringi_1.2.4            lattice_0.20-38          trimcluster_0.1-2.1      Matrix_1.2-15           
 [86] KMsurv_0.1-5             pillar_1.3.1             BiocManager_1.30.4       data.table_1.12.0        cowplot_0.9.4           
 [91] bitops_1.0-6             R6_2.3.0                 gridExtra_2.3            vipor_0.4.5              codetools_0.2-16        
 [96] assertthat_0.2.0         rhdf5_2.26.2             pkgmaker_0.27            withr_2.1.2              GenomeInfoDbData_1.2.0  
[101] hms_0.4.2                diptest_0.75-7           grid_3.5.2               tidyr_0.8.2              class_7.3-15            
[106] DelayedMatrixStats_1.4.0 rvcheck_0.1.3       
